test/suite/test_compact16.py

```diff
@@ -0,0 +1,93 @@
+#
+# In jurisdictions that recognize copyright laws, the author or authors
+# of this software dedicate any and all copyright interest in the
+# software to the public domain. We make this dedication for the benefit
+# of the public at large and to the detriment of our heirs and
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import threading, time
+from compact_util import compact_util
+from wiredtiger import stat
+from wtthread import checkpoint_thread
+
+# test_compact16
+# Test compact and checkpoint running concurrently, ensuring space is reclaimed after the compaction
+# call.
+class test_compact16(compact_util):
+    create_params = 'key_format=i,value_format=S,allocation_size=4KB,leaf_page_max=32KB,leaf_value_max=16MB'
+    conn_config = 'cache_size=100MB,statistics=(all),verbose=[compact:2]'
+    uri = 'table:test_compact16'
+
+    table_numkv = 1000 * 1000
+
+    def test_compact16(self):
+        # FIXME-WT-11399
+        if self.runningHook('tiered'):
+            self.skipTest("this test does not yet work with tiered storage")
+
+        # Create and populate a table.
+        self.session.create(self.uri, self.create_params)
+        self.populate(self.uri, 0, self.table_numkv)
+
+        # Write to disk.
+        self.session.checkpoint()
+
+        # Remove 1/4 of the data
+        c = self.session.open_cursor(self.uri, None)
+        for i in range(self.table_numkv):
+            if (i % 4 == 0):
+                c.set_key(i)
+                c.remove()
+        c.close()
+
+        self.delete_range(self.uri, self.table_numkv // 4)
+
+        self.reopen_conn()
+
+        # Run compact concurrently with another thread that continually creates checkpoints.
+        done = threading.Event()
+        ckpt = checkpoint_thread(self.conn, done)
+        try:
+            ckpt.start()
+
+            # Wait for checkpoint to start before calling compact.
+            ckpt_started = 0
+            while not ckpt_started:
+                stat_cursor = self.session.open_cursor('statistics:', None, None)
+                ckpt_started = stat_cursor[stat.conn.checkpoint_state][2] != 0
+                stat_cursor.close()
+                time.sleep(0.1)
+            self.session.compact(self.uri)
+        finally:
+            done.set()
+            ckpt.join()
+
+        # Compact should not leave more than 20% available space in the file. If it does, that means
```

Comment:
Can we also check we are hitting `session_table_compact_conflicting_checkpoint`?

Reply:
Not sure if this is a good idea in the end as we may not hit the race at all times.

Reply:
I would prefer to ignore test cases where we happen to hit it. So I don't think we should include that check.

Reply:
Sounds good